File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/stored/library/items/files/job/specs/GivenAFileThatExists/AndAnImpossibleFileRead/WhenProcessingTheJob.java
package com.lasthopesoftware.bluewater.client.stored.library.items.files.job.specs.GivenAFileThatExists.AndAnImpossibleFileRead;

import com.lasthopesoftware.bluewater.client.browsing.library.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.IStoredFileAccess;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJob;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJobProcessor;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJobState;
import com.lasthopesoftware.bluewater.client.stored.library.items.files.repository.StoredFile;
import com.lasthopesoftware.storage.read.permissions.IFileReadPossibleArbitrator;
import com.lasthopesoftware.storage.write.permissions.IFileWritePossibleArbitrator;
import com.namehillsoftware.handoff.promises.Promise;

import org.junit.BeforeClass;
import org.junit.Test;

import java.io.File;
import java.util.Collections;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class WhenProcessingTheJob {

	private static List<StoredFileJobState> jobStates;

	@BeforeClass
	public static void before() {
		final StoredFileJobProcessor storedFileJobProcessor = new StoredFileJobProcessor(
			storedFile -> {
				final File mockFile = mock(File.class);
				when(mockFile.exists()).thenReturn(true);
				return mockFile;
			},
			mock(IStoredFileAccess.class),
			(libraryId, f) -> Promise.empty(),
			mock(IFileReadPossibleArbitrator.class),
			mock(IFileWritePossibleArbitrator.class),
			(is, f) -> {});

		jobStates = storedFileJobProcessor.observeStoredFileDownload(
			Collections.singleton(new StoredFileJob(
				new LibraryId(12),
				new ServiceFile(1),
				new StoredFile(new LibraryId(12), 1, new ServiceFile(1), "test-path", true))))
			.map(j -> j.storedFileJobState)
			.toList()
			.blockingGet();
	}

	@Test
	public void thenTheFileStateIsUnreadable() {
		assertThat(jobStates).containsExactly(
			StoredFileJobState.Queued,
			StoredFileJobState.Unreadable);
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/stored/library/items/files/job/GivenAFileThatExists/AndAFileThatIsAlreadyDownloaded/WhenProcessingTheJob.kt
package com.lasthopesoftware.bluewater.client.stored.library.items.files.job.GivenAFileThatExists.AndAFileThatIsAlreadyDownloaded

import com.lasthopesoftware.bluewater.client.browsing.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJob
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJobProcessor
import com.lasthopesoftware.bluewater.client.stored.library.items.files.job.StoredFileJobState
import com.lasthopesoftware.bluewater.client.stored.library.items.files.repository.StoredFile
import com.namehillsoftware.handoff.promises.Promise
import io.mockk.every
import io.mockk.mockk
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.net.URI

class WhenProcessingTheJob {
	private val storedFileJobStatus by lazy {
		val storedFile = StoredFile(LibraryId(10), ServiceFile(1), URI("test-path"), true)
		storedFile.setIsDownloadComplete(true)
		val storedFileJobProcessor = StoredFileJobProcessor(
			mockk {
				every { promiseOutputStream(any()) } returns Promise.empty()
			},
			mockk { every { promiseDownload(any(), any()) } returns Promise.empty() },
			mockk(),
		)

		storedFileJobProcessor
			.observeStoredFileDownload(
				setOf(
					StoredFileJob(
						LibraryId(10),
						ServiceFile(1),
						storedFile
					)
				)
			)
			.map { f -> f.storedFileJobState }
			.toList().blockingGet()
	}

	@Test
	fun `then an already exists result is returned`() {
		assertThat(storedFileJobStatus)
			.containsExactly(StoredFileJobState.Queued, StoredFileJobState.Downloaded)
	}
}
